text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,154 @@
+cross-compiling. In the case where I wasn't, I couldn't complete the build because `rustc`
+complained about the libstd I was building overlapping with the libstd in the sysroot.
+
+For these reasons, most freestanding projects I know of avoid Cargo altogether, and just include
+submodule rust and run make in that. Cargo can still be used if one manages to get the requisite
+libraries in the sysroot. But this is a tedious operation that individual projects shouldn't need to
+reimplement, and one that has serious security implications if the normal libstd is modified.
+
+The fundamental plan proposed in this RFC is to make sure that anything Cargo builds never blindly
+links against libraries in the sysroot. This is achieved by making Cargo aware of all dependencies,
+including those libstd or its backing crates. That way, these problems are avoided.
+
+For the record, I first raised this issue [here](https://github.com/rust-lang/Cargo/issues/1096).
+
+
+# Detailed design
+
+The only new interface proposed is a boolean field in `Cargo.toml` specifying that the package does
+not depend on libstd by default. Note that this is technically orthogonal to Rust's `no_std`, as one
+might want to `use` their own build of libstd by default, or implicitly depend on it but not
+glob-import the prelude. To disambiguate, this field is called `implicit-deps`; please, go ahead and
+bikeshead the name. `implicit-deps` is true by default to maintain compatibility with existing
+packages. When true, "std" will be implicitly appended to the list of dependencies.
+
+When Cargo sees a package name it cannot resolve, it will query `rustc` for the default sysroot, and
+look inside to see if it can find a matching rlib. [It is necessary to query `rustc` because the
+`rustc` directory layout is not stabilized and `rustc` and Cargo are versioned independently. The
+same version issues make giving a Cargo a whitelist of potential standard library crate-names
+risky.] If a matching rlib is successful found, Cargo will copy it (or simlink it) into the
+project's build directly as if it built the rlib. Each rlib in the sysroot must be paired with some
+sort of manifest listing its dependencies, so Cargo can copy those too.
+
+`rustc` will have a new `--use-sysroot=<true|false>` flag. When Cargo builds a package, it will
+always pass `--use-sysroot=false` to `rustc`, as any rlibs it needs will have been copied to the
+build directory. Cargo can and will then pass those rlibs directly just as it does with normal Cargo
+deps.
+
+If Cargo cannot find the libraries it needs in the sysroot, or a library's dependency manifest is
+missing, it will complain that the standard libraries needed for the current job are missing and
+give up.
+
+## Future Compatibility
+
+In the future, rather than giving up if libraries are missing Cargo could attempt to download them
+from some build cache. In the farther future, the stdlib libraries may be Cargoized, and Cargo able
```

Comment:
Hmm, we over at NixOS maintain a build cache for Haskell and it works. If you only focus the slower release channels, and prioritize popular packages, it can still be useful.

The idea of a stable ABI scares me, but if/when it happens, that problem goes away too.